src/dynamic-launcher.c

```diff
@@ -0,0 +1,868 @@
+      return TRUE;
+    }
+
+  /* Do some validation on the icon before saving it */
+  icon_v = g_variant_get_variant (arg_icon_v);
+  if (!icon_v || !xdp_validate_serialized_icon (icon_v, TRUE /* bytes_only */))
+    {
+      g_set_error (&error,
+                   XDG_DESKTOP_PORTAL_ERROR, XDG_DESKTOP_PORTAL_ERROR_INVALID_ARGUMENT,
+                   "Dynamic launcher icon failed validation");
+      g_dbus_method_invocation_return_gerror (invocation, error);
+      return TRUE;
+    }
+
+  launcher_data = g_variant_new ("(svs)", arg_name, icon_v, arg_icon_format);
+  token = g_uuid_string_random ();
+
+  /* Save the token in memory and return it to the caller */
+  set_launcher_data_for_token (token, launcher_data);
+
+  xdp_dynamic_launcher_complete_request_install_token (object, invocation, token);
+
+  return TRUE;
+}
+
+static gboolean
+handle_uninstall (XdpDynamicLauncher    *object,
+                  GDBusMethodInvocation *invocation,
+                  const gchar           *arg_desktop_file_id,
+                  GVariant              *arg_options)
+{
+  Call *call = call_from_invocation (invocation);
+  const char *app_id = xdp_app_info_get_id (call->app_info);
+  g_autoptr(GError) error = NULL;
+  g_autofree char *icon_dir = NULL;
+  g_autofree char *icon_path = NULL;
+  g_autofree char *desktop_dir = NULL;
+  g_autoptr(GFile) icon_file = NULL;
+  g_autoptr(GFile) desktop_file = NULL;
+  g_autoptr(GFile) link_file = NULL;
+  g_autoptr(GKeyFile) desktop_keyfile = NULL;
+
+  if (!validate_desktop_file_id (app_id, arg_desktop_file_id, &error))
+    {
+      g_dbus_method_invocation_return_gerror (invocation, error);
+      return TRUE;
+    }
+
+  icon_dir = g_build_filename (g_get_user_data_dir (), "xdg-desktop-portal", "icons", NULL);
+  desktop_dir = g_build_filename (g_get_user_data_dir (), "xdg-desktop-portal", "applications", NULL);
+  g_mkdir_with_parents (desktop_dir, 0700);
+  g_mkdir_with_parents (icon_dir, 0700);
+
+  link_file = g_file_new_build_filename (g_get_user_data_dir (), "applications", arg_desktop_file_id, NULL);
+  if (!g_file_delete (link_file, NULL, &error))
+    {
+      if (g_error_matches (error, G_IO_ERROR, G_IO_ERROR_NOT_FOUND))
+        g_info ("Uninstall() method failed because launcher '%s' does not exist", arg_desktop_file_id);
+      goto error;
+    }
+
+  desktop_file = g_file_new_build_filename (desktop_dir, arg_desktop_file_id, NULL);
+  desktop_keyfile = g_key_file_new ();
+  if (g_key_file_load_from_file (desktop_keyfile, g_file_peek_path (desktop_file), G_KEY_FILE_NONE, NULL))
+    icon_path = g_key_file_get_string (desktop_keyfile, G_KEY_FILE_DESKTOP_GROUP, "Icon", NULL);
+
+  if (!g_file_delete (desktop_file, NULL, &error))
+    goto error;
+
+  if (icon_path && g_str_has_prefix (icon_path, icon_dir))
+    {
+      icon_file = g_file_new_for_path (icon_path);
+      if (!g_file_delete (icon_file, NULL, &error))
+        goto error;
+    }
```

Comment:
I wonder if this particular function should be a bit more lenient with errors. For example, here, if deleting the desktop file errs, the icon won't be deleted. I think it should try and delete all files, even when one file errs on deletion.

Reply:
Makes sense